types/wlr_input_method_v2.c

```diff
@@ -101,10 +119,177 @@ static void im_get_input_popup_surface(struct wl_client *client,
+	struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab =
+		wl_container_of(listener, im_keyboard_grab, keymap_listener);
+	im_keyboard_grab_send_keymap(im_keyboard_grab, data);
+}
+
+static void handle_keyboard_repeat_info(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab =
+		wl_container_of(listener, im_keyboard_grab, keymap_listener);
+	im_keyboard_grab_send_repeat_info(im_keyboard_grab, data);
+}
+
+void wlr_input_method_keyboard_grab_v2_set_keyboard(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	if (keyboard == im_keyboard_grab->keyboard) {
+		return;
+	}
+
+	if (im_keyboard_grab->keyboard) {
+		wl_list_remove(&im_keyboard_grab->keymap_listener.link);
+		wl_list_remove(&im_keyboard_grab->repeat_info_listener.link);
+	}
+
+	if (im_keyboard_grab->keyboard == NULL ||
+			strcmp(im_keyboard_grab->keyboard->keymap_string,
+			keyboard->keymap_string) != 0) {
+		// send keymap only if it is changed, or if input method is not
+		// aware that it did not change and blindly send it back with
+		// virtual keyboard, it may cause an infinite recursion.
+		if (!im_keyboard_grab_send_keymap(im_keyboard_grab, keyboard)) {
+			wlr_log(WLR_ERROR, "Failed to send keymap for input-method keyboard grab");
+			return;
+		}
+	}
+	im_keyboard_grab_send_repeat_info(im_keyboard_grab, keyboard);
+
+	im_keyboard_grab->keyboard = keyboard;
+	im_keyboard_grab->keymap_listener.notify = handle_keyboard_keymap;
+	wl_signal_add(&keyboard->events.keymap,
+		&im_keyboard_grab->keymap_listener);
+	im_keyboard_grab->repeat_info_listener.notify =
+		handle_keyboard_repeat_info;
+	wl_signal_add(&keyboard->events.repeat_info,
+		&im_keyboard_grab->repeat_info_listener);
+};
 
 static void im_grab_keyboard(struct wl_client *client,
 		struct wl_resource *resource, uint32_t keyboard) {
-	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::grab_keyboard");
+	struct wlr_input_method_v2 *input_method =
+		input_method_from_resource(resource);
+	if (!input_method) {
+		return;
+	}
+	if (input_method->im_keyboard_grab) {
+		// Already grabbed
+		return;
+	}
+	struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab =
+		calloc(1, sizeof(struct wlr_input_method_keyboard_grab_v2));
+	if (!im_keyboard_grab) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	struct wl_resource *im_keyboard_grab_resource = wl_resource_create(
+		client, &zwp_input_method_keyboard_grab_v2_interface,
+		wl_resource_get_version(resource), keyboard);
+	if (im_keyboard_grab_resource == NULL) {
+		free(im_keyboard_grab);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(im_keyboard_grab_resource,
+		&im_keyboard_grab_impl, im_keyboard_grab,
+		im_keyboard_grab_resource_destroy);
+	im_keyboard_grab->resource = im_keyboard_grab_resource;
+	im_keyboard_grab->input_method = input_method;
+	im_keyboard_grab->seat_client = wlr_seat_client_for_wl_client(
```

Comment:
The `input_method` is already tied to a seat. Why do we need this `seat_client` field?

Reply:
Ah, I suppose this is to get seat serials.